Fix issue Montonically increasing AssemblyFileVersion to satisfy the …

ruhullahshah opened the pull request:
Fix for issue https://github.com/GitTools/GitVersion/issues/1366

Implementation methodology same as suggested by @jbaehr in https://github.com/GitTools/GitVersion/issues/1157

jbaehr commented:
Thanks a lot for your work. I gave it a try and it works like a charm (just remember to call `GitVersion.exe` with `-nocache` when playing around ;-) ).

I would love a comment from the maintainers whether this approach has chances to get merged. @asbjornu, you already gave thumbs up on my comment in #1157... Please tell us if you request further changes to the implementation, format string syntax, or the like.

ruhullahshah commented:
@jbaehr You are welcome and thanks for testing out the changes :)

asbjornu commented on src/GitVersionCore.Tests/StringFormatWithExtensionTests.cs:
```diff
@@ -0,0 +1,75 @@
+        {
+            Environment.SetEnvironmentVariable("GIT_VERSION_TEST_VAR", "Env Var Value");
+            var propertyObject = new { };
+            var target = "{$GIT_VERSION_TEST_VAR}";
```
To avoid collisions, I would prefer if we could use a more explicit prefix here, like `env:`.

ruhullahshah replied:
Sure we could but I think from the reaction on approach suggested by @jbaehr on https://github.com/GitTools/GitVersion/issues/1157 we agreed to use $ as the prefix for enviroment variables and I am not sure as to what the collisions would be here since the usage of $ occurs in the format string and we parse them internally.

asbjornu replied:
The collision I'm thinking of is between non-environment variables and GitVersion's variables. We might introduce system variables and whatnot in the future, so having a prefix that clearly identifies where the variable is coming from is more future proof and intuitive than just `$` imho.

jbaehr replied:
Using `$` for dereferencing env vars was my initial proposal as for me it is very intuitive --
 it's was bash is using.
Anyway, I have no strong opinion on that; I'm also fine with `env:`.

ruhullahshah replied:
Addressed in the latest commit

asbjornu commented on src/GitVersionCore.Tests/StringFormatWithExtensionTests.cs:
```diff
@@ -0,0 +1,75 @@
+        public void FormatWith_MultipleTokensAndVerbatimText()
+        {
+            var propertyObject = new { SomeProperty = "SomeValue", AnotherProperty = "Other Value" };
+            var target = "{SomeProperty} some text {AnotherProperty}";
```
Will these "bare" format strings work for the variables provided by GitVersion? Perhaps we could make that explicit with a `gv:` prefix?

ruhullahshah replied:
This testcase just tests the existing functionality and specifically tests the `FormatWith` method. I guess this should not be a concern

jbaehr replied:
Note that introducing a "gv:"-prefix for git version properties would break existing config files. It may be acceptable to introduce breaking changes while still in beta phase, but that is something I'd rather do in a dedicated PR.

asbjornu replied:
True, scrap that. No prefix means GitVersion's own variables.

asbjornu commented on src/GitVersionCore/Configuration/Config.cs:
```diff
@@ -26,6 +26,9 @@ public Config()
         public string AssemblyInformationalFormat { get; set; }
 
+        [YamlMember(Alias = "assembly-file-versioning-format")]
+        public string AssemblyFileVersioningFormat { get; set; }
```
Shouldn't we provide `AssemblyVersioningFormat` as well?

ruhullahshah replied:
Addressed in the latest patch with some code cleanup

asbjornu commented on src/GitVersionCore/Helpers/EnvironmentHelper.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        public static string GetEnvironmentVariableForProcess(string envVar)
+        {
+            return Environment.GetEnvironmentVariable(envVar, EnvironmentVariableTarget.Process);
```
Could we make an abstraction on top of `Environment` here so it's translated to something less specific like `IDictionary<string, string>`? That would make testing easier and the code more portable.

ruhullahshah replied:
I am not sure exactly as to how could I improve here and provide a useful abstraction.  Seems like Mono https://github.com/mono/mono/blob/master/mcs/class/corlib/System/Environment.cs uses the same approach as we have. In case there are multiple env vars, a dictionary would make sense but in this case we seek the value of a single env var and this should be fine

asbjornu replied:
I just want a translation so our "core" isn't dependent on the `System.Environment` class itself. Perhaps `Func<string, string>` is a better abstraction?

jbaehr replied:
Using `Func<string, string>` has the advantage that one can use `System.Environment.GetEnvironmentVariable` (or even the currently used `Helpers.EnvironmentHelper.GetEnvironmentVariableForProcess`) directly at the call site.

What is your preference here? Turning this formatter into a class and use ctor injection or rather give the currently used string extension method an additional parameter?
(e.g. `public static string FormatWith<T>(this string template, T source, Func<string, string> getEnvironmentVariable)`)

ruhullahshah replied:
`GitVersionCore\Authentication.cs`,  `GitVersionCore\BuildServers\*.cs` etc all use a naked call to `System.Environment.GetEnvironmentVariable`. How exactly does my patch create a problem when there are already existing reference to System.Environment in the GitVersionCore? Maybe abstracting all the System.Environment references in the GitVersionCore can be a part of a cleanup but I am not sure how drastically does it change things here.

asbjornu replied:
True, I did not think of the existing calls to `System.Environment`. I agree; let's address that abstraction in a future PR.

asbjornu commented on src/GitVersionCore/StringFormatWith.cs:
```diff
@@ -8,7 +8,7 @@ namespace GitVersion
     static class StringFormatWithExtension
     {
-        private static readonly Regex TokensRegex = new Regex(@"{\w+}", RegexOptions.Compiled);
+        private static readonly Regex TokensRegex = new Regex(@"{\$??\w+(\?\?\w+)??}", RegexOptions.Compiled);
```
Since this is based on a regex, I would love to see some negative tests as well that challenges the regex engine a bit, just to make sure we don't have a too greedy expression or something that causes performance problems.

ruhullahshah replied:
Addressed with some additional testcases

ruhullahshah commented:
@asbjornu You are welcome. I have addressed some of your review comments and provided my opinion on others. In my opinion, we have the functionality in place that addresses the theme of this issue.

asbjornu commented on src/GitVersionCore.Tests/StringFormatWithExtensionTests.cs:
```diff
@@ -0,0 +1,119 @@
+        {
+            Environment.SetEnvironmentVariable("GIT_VERSION_TEST_VAR", "Env Var Value");
+            var propertyObject = new { };
+            var target = "{env:GIT_VERSION_TEST_VAR??fallback}";
```
Are spaces allowed around the null propagation operator? That would be a readability improvement, imho. Is that hard to add?

ruhullahshah replied:
Addressed in the latest patch.

ruhullahshah commented:
You are welcome. I have included the changes for the null propagation operator readability and updated some test cases.

asbjornu commented on src/GitVersionCore/OutputVariables/VariableProvider.cs:
```diff
@@ -100,5 +94,24 @@ static void PromoteNumberOfCommitsToTagNumber(SemanticVersion semanticVersion)
             semanticVersion.BuildMetaData.CommitsSinceTag = null;
         }
+
+        static void CheckAndFormatString<T>(string formatString, T source,  string defaultValue, string formatVarName, out string formattedString)
```
Neat, I like this. But is there a reason why the `formattedString` is an `out` parameter instead of a return value?

ruhullahshah replied:
No preferences. Matter of taste

ruhullahshah replied:
Addressed in the latest commit

asbjornu commented on src/GitVersionCore.Tests/StringFormatWithExtensionTests.cs:
```diff
@@ -0,0 +1,129 @@
+        {
+            Environment.SetEnvironmentVariable("GIT_VERSION_TEST_VAR", "Env Var Value");
+            var propertyObject = new { };
+            var target = "{env:GIT_VERSION_TEST_VAR ?? fallback}";
```
👍

asbjornu commented on src/GitVersionCore/StringFormatWith.cs:
```diff
@@ -27,16 +27,44 @@ public static string FormatWith<T>(this string template, T source)
 
             // {MajorMinorPatch}+{Branch}
             var objType = source.GetType();
+            const string ENV_VAR_INDICATOR = "env:";
```
Perhaps this `const` can be moved up alongside the `TokensRegex`?

ruhullahshah replied:
Addressed in the latest commit.

asbjornu commented on src/GitVersionCore/StringFormatWith.cs:
```diff
@@ -27,16 +27,44 @@ public static string FormatWith<T>(this string template, T source)
+                // Support evaluation of environment variables in the format string
+                // For example: {env:JENKINS_BUILD_NUMBER??fall-back-string}
+
+                if (memberAccessExpression.StartsWith(ENV_VAR_INDICATOR))
```
Can't you just check for the match of `env:` in the regex instead of using this constant? With [Named Matched Subexpressions](https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#named-matched-subexpressions) in the regex you can reference the [match group](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.group?view=netframework-4.7.1) by name and check its [`.Success` property](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.group.success).

ruhullahshah replied:
Addressed in the latest commit

ruhullahshah commented:
@asbjornu I would like to address all the changes you request in one go. This is getting repetitive. The review before this should have had these comments because these changes have been there before the last commit. I will give you some time to make up your mind about the PR as a whole and once you have reviewed all the code let me know, i will address your changes.

asbjornu commented:
@ruhullahshah: Sorry about that. I was sure the latest review was for new code. I should of course have reviewed that in the first go, but if I remember correctly, I had limited time when I did the first review so it wasn't as thorough as I would have liked.

asbjornu commented:
Excellent. Thanks for this!

ruhullahshah commented:
You are welcome :)

vishalvatsal commented:
Hi Team,

I have a need to use build number in the assembly version for my TeamCity build. 

Following is my gitversion config

`next-version: 1.0.0
assembly-versioning-scheme: MajorMinorPatch
assembly-versioning-format: '{Major}.{Minor}.{Patch}.{env:BUILD_COUNTER??0}'
assembly-file-versioning-scheme: MajorMinorPatch
continuous-delivery-fallback-tag: ''
mode: ContinuousDeployment
branches:
    release:
        tag: ''        
    develop:
        tag: 'dev'
    master:
        tag: 'master'
ignore:
  sha: []
merge-message-formats: {}`


In Team City project I have an environment variable

`Name: env.BUILD_COUNTER
 Value: %build.counter%'

After the build this is what I get
**GitVersion.AssemblySemVer**  =  **1.0.0.{env:BUILD_COUNTER??0}**


Can you please let me know what I am doing wrong here? Or what do I need to do to make sure env:BUILD_COUNTER gets substituted with the correct value?

Thank you

jbaehr commented:
@vishalvatsal The implementation in this PR explicitly requires spaces around the `??` in order to treat it as a fallback-separator. Your config sample seem to omit those spaces.

Note however, that the currently latest GitVersion release features an enhanced implementation, that further allows fallback strings for regular GitVersion variables, too, and also allows for quoted fallback strings containing special characters. cf. #2179

vishalvatsal commented:
Thanks a ton @jbaehr.

It works now.
